### PR TITLE
CDAP-19023: add support for configuring composite trigger

### PR DESCRIPTION
**Pull request by de-lan**
This PR adds support for configuring composite trigger, more specifically it adds the id information for the respective composite trigger. Originally schedule trigger was default to the PROGRAM_STATUS type when parsing runtime arguments. The sister UI pr is here: https://github.com/cdapio/cdap-ui/pull/414

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java [hunk 1]
```diff
@@ -381,6 +383,97 @@ private PipelinePlan createPlan() {
     return planner.plan(spec);
   }
 
+  private void updateTokenWithTriggeringCompositeProperties(TriggeringScheduleInfo scheduleInfo,
```
Comment (de-lan) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java]: This function behaves exactly like `updateTokenWithTriggeringProperties`, except that it iterates through all the triggers and match the id information

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java [hunk 2]
```diff
@@ -48,11 +59,21 @@ public String getTarget() {
...
+    return pipelineId;
+  }
+
   @Override
   public String toString() {
     return "ArgumentMapping{" +
       "source='" + getSource() + '\'' +
       ", target='" + getTarget() + '\'' +
+        ", triggeringPipelinePropertyId='" +
+        (pipelineId == null ? "" : pipelineId.toString()) +  '\'' +
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: don't need to do this, just do `+ pipelineId` like it's done for source and target.

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java [hunk 3]
```diff
@@ -48,11 +59,21 @@ public String getTarget() {
...
+  public TriggeringPipelineId getTriggeringPipelineId() {
+    return pipelineId;
+  }
+
   @Override
   public String toString() {
     return "ArgumentMapping{" +
       "source='" + getSource() + '\'' +
       ", target='" + getTarget() + '\'' +
+        ", triggeringPipelinePropertyId='" +
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: indentation should be 2 spaces, make sure your IDE has the right settings.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java [hunk 4]
```diff
@@ -625,8 +626,17 @@ private WorkflowManager deployPipelineWithSchedule(String pipelineName, Engine e
 
     String defaultNamespace = NamespaceId.DEFAULT.getNamespace();
     // Use properties from the triggering pipeline as values for runtime argument key1, key2
+    ArgumentMapping key1MappingWithId = new ArgumentMapping(
+        key1Mapping.getSource(),
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java]: indentation

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java [hunk 5]
```diff
@@ -625,8 +626,17 @@ private WorkflowManager deployPipelineWithSchedule(String pipelineName, Engine e
...
+        key1Mapping.getTarget(),
+        new TriggeringPipelineId(defaultNamespace, triggeringPipelineName));
+    PluginPropertyMapping key2MappingWithId = new PluginPropertyMapping(
+        key2Mapping.getStageName(),
+        key2Mapping.getSource(),
+        key2Mapping.getTarget(),
+        new TriggeringPipelineId(defaultNamespace, triggeringPipelineName));
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java]: this doesn't really test that the logic is correct. Should have a test where both pipelines have the same property name in order to make sure the triggering pipeline name is taken into account correctly

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/PluginPropertyMapping.java [hunk 6]
```diff
@@ -43,7 +51,9 @@ public String toString() {
     return "PluginPropertyMapping{" +
       "source='" + getSource() + '\'' +
       ", target='" + getTarget() + '\'' +
-      "stageName='" + getStageName() + '\'' +
+      ", stageName='" + getStageName() + '\'' +
+        ", triggeringPipelinePropertyId='" +
+        (getTriggeringPipelineId() == null ? "" : getTriggeringPipelineId().toString()) + '\'' +
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/PluginPropertyMapping.java]: similar comment, don't need the null check

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/PluginPropertyMapping.java [hunk 7]
```diff
@@ -30,6 +30,14 @@ public PluginPropertyMapping(@Nullable String stageName, @Nullable String source
     this.stageName = stageName;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/PluginPropertyMapping.java]: this should be updated to call the other constructor

```
this(stageName, source, target, null)
```

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  public String getPipelineName() {
+    return pipelineName;
+  }
+
+  public boolean equals(TriggeringPipelineId otherPipelineId) {
+    if (otherPipelineId == null) {
+      return false;
+    }
+    return namespace.equals(otherPipelineId.getNamespace())
+        && pipelineName.equals(otherPipelineId.getPipelineName());
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java]: indentation

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  @Override
+  public String toString() {
+    return "TriggeringPipelinePropertyId{" +
+        "namespace='" + getNamespace() + '\'' +
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java]: indentation

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java (hunk 1 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java]: should not copy-paste code like this. Refactor things so that this method calls updateTokenWithTriggeringProperties.

While you're at it, I think it would also be a good idea to change updateTokenWithTriggeringProperties so that it is a static method that returns a Map of keys rather than mutating the WorkflowToken within the method. That way you can add unit tests that calls the method directly rather than the current tests that need to deploy and run entire pipelines

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java (hunk 2 above)
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: `pipelineId` is a Class instead of String like source and target.  Guess have to do a null check...

Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: if it has a toString() method java will use that, you don't need to do this manually

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java (hunk 5 above)
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java]: will update with a new dedicated test

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java (hunk 2 above)
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: gotcha! that's a good point, thanks!

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java (hunk 1 above)
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java]: Thanks! updated

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java [hunk 10]
```diff
@@ -114,6 +112,8 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 
+import static io.cdap.cdap.features.Feature.PIPELINE_COMPOSITE_TRIGGERS;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java]: style: for readability, we don't do static imports.

Import just the Feature class and refer to it as Feature.PIPELINE_COMPOSITE_TRIGGERS

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+import io.cdap.cdap.etl.proto.v2.PluginPropertyMapping;
+import io.cdap.cdap.etl.proto.v2.TriggeringPipelineId;
+import io.cdap.cdap.etl.proto.v2.TriggeringPropertyMapping;
+import io.cdap.cdap.internal.io.SchemaTypeAdapter;
+import org.slf4j.Logger;
+
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Map;
+
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: should have a short javadoc for the class

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+public final class WorkflowScheduleUtils {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: nit: let's name it PipelineTriggers so that it's more specific.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+   *
+   * @param triggerInfo {@link ProgramStatusTriggerInfo} with plugin properties
+   * @param propertiesMapping  {@link TriggeringPropertyMapping} for using the plugin
+   * @param logger  {@link Logger} the logger
+   * @param compositeTriggerEnabled  whether the composite trigger is enabled
+   */
+  public static void processSchedulePropertiesMapping(Map<String, String> mappings,
+                                                      ProgramStatusTriggerInfo triggerInfo,
+                                                      TriggeringPropertyMapping propertiesMapping,
+                                                      Logger logger,
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: don't pass in a Logger, create a static Logger for this class like in other places. Log messages should be come from a Logger for the class that has the code.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+    TriggeringPipelineId pipelineId = new TriggeringPipelineId(
+      triggerInfo.getNamespace(),
+      triggerInfo.getApplicationName());
+
+    BasicArguments triggeringArguments = new BasicArguments(
+      triggerInfo.getWorkflowToken(),
+      triggerInfo.getRuntimeArguments());
+
+    // Get the value of every triggering pipeline arguments specified in the propertiesMapping and update newRuntimeArgs
+    List<ArgumentMapping> argumentMappings = propertiesMapping.getArguments();
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: is this just copy-pasted code? It's a bit hard to follow, so if there are any differences from before can you comment on them?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+  /**
+   * Parse the arguments mapping from the pipeline schedule.
+   *
+   * @param triggerInfo {@link ProgramStatusTriggerInfo} with plugin properties
+   * @param propertiesMapping  {@link TriggeringPropertyMapping} for using the plugin
+   * @param logger  {@link Logger} the logger
+   * @param compositeTriggerEnabled  whether the composite trigger is enabled
+   */
+  public static void processSchedulePropertiesMapping(Map<String, String> mappings,
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: nit: process -> add

This makes it a little more clear that it is adding things to the map.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .registerTypeAdapter(FieldOperation.class, new FieldOperationTypeAdapter()).create();
+  private static final String RESOLVED_PLUGIN_PROPERTIES_MAP = "resolved.plugin.properties.map";
+  private static final Type STAGE_PROPERTIES_MAP = new TypeToken<Map<String, Map<String, String>>>() { }.getType();
+
+  private WorkflowScheduleUtils() {
+  }
+
+  /**
+   * Parse the arguments mapping from the pipeline schedule.
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: this description should have more detail. It should include some example input and what gets added to the mappings.

Somebody should be able to read this and tell what the method does, especially since the code isn't the clearest.

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java [hunk 17]
```diff
@@ -26,10 +26,21 @@ public class ArgumentMapping {
   private final String source;
   @Nullable
   private final String target;
+  @Nullable
+  private final TriggeringPipelineId pipelineId;
 
   public ArgumentMapping(@Nullable String source, @Nullable String target) {
     this.source = source;
     this.target = target;
+    this.pipelineId = null;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: change this to call `this(source, target, null)`

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+ */
+
+package io.cdap.cdap.etl.proto.v2;
+
+/**
+ * Class for identifiers of properties from the triggering pipeline.
+ */
+public class TriggeringPipelineId {
+  private final String namespace;
+  private final String pipelineName;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java]: pipelineName is redundant, can rename this to just `name`.

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java [hunk 19]
```diff
@@ -26,10 +26,21 @@ public class ArgumentMapping {
   private final String source;
   @Nullable
   private final String target;
+  @Nullable
+  private final TriggeringPipelineId pipelineId;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: rename this to just `pipeline`. It's a little more future-proof in case we ever need to add more fields to the TriggeringPipelineId object and it no longer is an id.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java (hunk 14 above)
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: yes, this is partially copied code, but also with some new code checking on the pipeline ID

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java (hunk 10 above)
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java]: updated

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java (hunk 11 above)
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: javadoc added

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java (hunk 19 above)
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: updaed

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+   * {"namespace":"default","pipelineName":"Pipeline2"}}]}" and a trigger {"namespace":"default","pipelineName":
+   * "Pipeline1"} which has runtime arguments "{"source1": "value1","source2":"value2"}". When composite pipeline
+   * trigger is enabled, only {"target1": "value1"} is added. Otherwise, the code will ignore the ID check and
+   * {"target1": "value1", "target2": "value2"} is added.
+   *
+   * @param triggerInfo {@link ProgramStatusTriggerInfo} with plugin properties
+   * @param propertiesMapping  {@link TriggeringPropertyMapping} for using the plugin
+   * @param compositeTriggerEnabled  whether the composite trigger is enabled
+   */
+  public static void addSchedulePropertiesMapping(Map<String, String> mappings,
```
Comment (de-lan) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java]: Most of the code here is from `SmartWorkflow.java`, with slight difference that now it checks the ID if `compositeTriggerEnabled` is true

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java [hunk 21]
```diff
@@ -31,34 +31,45 @@
...
 
 import java.lang.reflect.Type;
 import java.util.List;
 import java.util.Map;
 
-public final class WorkflowScheduleUtils {
+/**
+ * Handles parsing of pipeline runtime arguments mapping.
+ */
+public final class PipelineTriggersUtils {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java]: nit: we normally don't put 'Utils' in the name, just PipelineTriggers.

**Comment by albertshau** [comment]
for future reference, you need to add the 'build' label in order to tell it to run the build

**Review by albertshau — APPROVED** [review]
lgtm assuming the build passes. Please squash commits before merging.

**Comment by de-lan** [comment]
> for future reference, you need to add the 'build' label in order to tell it to run the build

gotcha, will do in future

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java (hunk 21 above)
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java]: Aha, I followed the other files naming convention, will update

**Comment by de-lan** [comment]
Mis-clicked the close button...
